### PR TITLE
AK+Everywhere: Make the Variant<Empty, …> pattern slightly easier

### PR DESCRIPTION
**Pull request by BenWiederhake**
I've noticed that we use `AK::Variant<Empty, …>` in several places, but didn't have any default constructor for it in that case, so we used to have to write `VariantWithEmpty m_foo { Empty {} };` every time, which is slightly silly.

Furthermore, it turns out that there was a use of `Variant<Empty, …>` that didn't even use the `Empty` option, to the point where that value would have been a bug.

### CODE REVIEW

File: AK/Variant.h [hunk 1]
```diff
@@ -210,7 +210,9 @@ struct Variant
     template<typename... NewTs>
     friend struct Variant;
 
-    Variant() = delete;
+    Variant() requires(!can_contain<Empty>()) = delete;
+    Variant()
+        : Variant(Empty()) {};
```
Comment (alimpfard) [inline AK/Variant.h:215]:
```suggestion
    Variant() requires(can_contain<Empty>())
        : Variant(Empty())
    {
    }
```

**Review by alimpfard — COMMENTED** [review]
tbh I'm not a big fan of this behaviour, you wouldn't expect `AK::Result<T, E>` to try default initialise an `E`, so why would you expect `Variant<Empty, ...>` to be initialised to...anything?

**Comment by BenWiederhake** [comment]
I don't see the parallel. It's a Variant that is meant to be possibly Empty, and it's default-initialized with nothing in it, so it's Empty. It's got nothing to do with a Result, and I agree that a default-empty Result would be confusing.

File: AK/Variant.h (hunk 1 above)
Comment (BenWiederhake) [reply AK/Variant.h:215]: Right!

**Comment by BenWiederhake** [comment]
Changes since first push:
- Adopt the `requires(can_contain<Empty>())` suggestion (no idea why I didn't do that in the first place)
- Not rebased since there weren't any commits to `master`. Huh.

**Comment by alimpfard** [comment]
I guess I'm just not very happy to add a "but" to "a variant must be explicitly initialised", which I suppose is where the similarity to Result comes from (at least, in my head)
I personally don't think making it default initialise to Empty is that much better than just explicitly initialising it.

Regardless, code is okay but I'm not convinced the extra bit of behind-the-curtain action is needed.

**Comment by BenWiederhake** [comment]
What can I say, I just disagree; we have default-initializers for other things, too, that have a meaningful "empty" state. String and Optional comes to mind, and a quick look at `AK/` shows also Bitmap, Checked, Complex, DisjointSpans (huh, didn't even know about that class!), FixedArray, FixedPoint, FlyString, Function (wtf?), IPv4Address, JsonValue, MACAddress, RefPtr, UFixedBigInt, URL, Utf{8,16,32}View, UUID (wtf?), WeakPtr, plus all the containers :sweat_smile:

I agree that maybe not all of these should be default-initializable (what a word!), but making a `Variant<Empty, …>` Empty by default seems just natural to me.

Let me know if that's still a hard-no from you, then I'll toss the change and only keep the "LibIMAP: Remove unused Variant option" commit.

**Comment by creator1creeper1** [comment]
Would it also make sense to add an `operator bool()` in this case so that we can do `if (variant)`?

**Comment by alimpfard** [comment]
> Let me know if that's still a hard-no from you

I'm not sure if I hate it enough for a hard no, so let's keep it and see where it takes us.

**Comment by alimpfard** [comment]
> Would it also make sense to add an `operator bool()`

`operator bool` is...honestly a big footgun, and what does `if (variant)` even mean?

- `if (auto p = variant.get_pointer<bool>(); *p)`?
- `if (variant.visit([](auto& v) -> bool { return v; }))`?
- `if (!variant.has<Empty>())`?
- something entirely different?

**Comment by creator1creeper1** [comment]
> * `if (!variant.has<Empty>())`?

I was thinking of this one but you're right, there are too many ambiguities.

**Comment by BenWiederhake** [comment]
For fun (and no profit) I implemented @creator1creeper1's suggestion here: https://github.com/BenWiederhake/serenity/commit/a3b5c133d8e8aa1327063126277c8d87799fba52

Looks reasonable enough to me, but I won't make it a PR.

**Pull request merged**